src/flight.rs

```diff
@@ -193,3 +202,183 @@ impl SqlFlightClient {
         Ok(stream)
     }
 }
+
+/// Represents the current state of the `RetryableQueryStream` state machine.
+/// Wraps a `FlightRecordBatchStream` and started from `Streaming` stage.
+/// If a retryable error occurs during streaming, the stream resets and retries.
+/// `Streaming` -> `Ready` → `Executing` → `Streaming` → `Ready`
+/// If a non-retryable error occurs during streaming, the stream will be immediately terminated.
+/// `Streaming` -> `Terminated`. (non-retryable error)
+enum StreamState {
+    /// Ready to retry a query
+    Ready,
+    /// Query is being executed, waiting for the server to return a stream
+    Executing(Pin<Box<dyn Future<Output = Result<FlightRecordBatchStream, GenericError>> + Send>>),
+    /// Initial state, actively streaming record batches from the server
+    Streaming(Pin<Box<FlightRecordBatchStream>>),
+    /// Terminal state - stream has ended due to non-retryable error
+    Terminated,
+}
+
+/// A retryable stream for executing SQL queries with Flight.
+///
+/// This stream automatically handles streaming failures and immediately retries queries.
+/// It yields `RecordBatch` results on success and `SpiceClientError` on failure.
+///
+/// ## Retry Behavior
+///
+/// When a connection reset occurs during streaming, the stream will:
+/// 1. Yield a `SpiceClientError::ConnectionReset` error to the consumer
+/// 2. If the consumer continues polling, automatically retry the entire query from the beginning
+/// 3. If the consumer stops polling, the stream will not retry and enters the `Terminated` state
+/// 4. Stop retrying and enters the `Terminated` state after reaching `max_retries` attempts
+///
+/// ## Consumer Options
+///
+/// **Option 1: Continue polling for automatic retry**
+/// ```text
+/// Poll 1: Ok(batch1)
+/// Poll 2: Ok(batch2)
+/// Poll 3: Err(ConnectionReset) → Consumer continues polling
+/// Poll 4: Ok(batch1) → Query restarted from beginning
+/// Poll 5: Ok(batch2)
+/// Poll 6: Ok(batch3)
+/// ...
+/// ```
+///
+/// **Option 2: Stop on error**
+/// ```text
+/// Poll 1: Ok(batch1)
+/// Poll 2: Ok(batch2)
+/// Poll 3: Err(ConnectionReset) → Consumer stops polling
+/// ```
+///
+/// ## Important Notes
+/// - The query restarts from the beginning on retry - previously yielded batches will be re-yielded
+/// - Non-retryable errors are returned immediately without retry attempts
+/// - Only connection resets and specific gRPC errors trigger retries
+///
+pub struct RetryableQueryStream {
+    client: Arc<SqlFlightClient>,
+    sql: Arc<String>,
+    params: Arc<Option<RecordBatch>>,
```

Comment:
```suggestion
    params: Option<RecordBatch>,
```

This doesn't need to be in an Arc, the record batch itself has Arc over the arrays.